pyomo/repn/linear_template.py

```diff
@@ -0,0 +1,358 @@
+                check_duplicates,
+            )
+            indent = '    ' * (len(subsets))
+            ans.extend(indent + line for line in subans)
+            constant += subconst
+        return ans, constant
+
+    def compile(
+        self,
+        env,
+        smap,
+        expr_cache,
+        args,
+        remove_fixed_vars=False,
+        check_duplicates=False,
+    ):
+        ans, constant = self._build_evaluator(
+            smap, expr_cache, 1, 1, remove_fixed_vars, check_duplicates
+        )
+        if not ans:
+            return constant
+        indent = '\n    '
+        if not constant and ans and ans[0].startswith('const +='):
+            # Convert initial "const +=" to "const ="
+            ans[0] = ''.join(ans[0].split('+', 1))
+        else:
+            ans.insert(0, 'const = ' + repr(constant))
+        fcn_body = indent.join(ans[1:])
+        if 'const' not in fcn_body:
+            # No constants in the expression.  Move the initial const
+            # term to the return value and avoid declaring the local
+            # variable
+            ans = ['return ' + ans[0].split('=', 1)[1]]
+            if fcn_body:
+                ans.insert(0, fcn_body)
+        else:
+            ans = [ans[0], fcn_body, 'return const']
+        if check_duplicates:
+            ans.insert(0, f"def build_expr(linear, {', '.join(args)}):")
+        else:
+            ans.insert(
+                0, f"def build_expr(linear_indices, linear_data, {', '.join(args)}):"
+            )
+        ans = indent.join(ans)
+        # build the function in the env namespace, then remove and
+        # return the compiled function.  The function's globals will
+        # still be bound to env
+        exec(ans, env)
+        return env.pop('build_expr')
+
+
+class LinearTemplateBeforeChildDispatcher(linear.LinearBeforeChildDispatcher):
+
+    def _before_indexed_var(self, visitor, child):
+        if child not in visitor.indexed_vars:
+            visitor.var_recorder.add(child)
+            visitor.indexed_vars.add(child)
+        return False, (_VARIABLE, child)
+
+    def _before_indexed_param(self, visitor, child):
+        if child not in visitor.indexed_params:
+            visitor.indexed_params.add(child)
+            name = visitor.symbolmap.getSymbol(child)
+            visitor.env[name] = child.extract_values()
+        return False, (_CONSTANT, child)
+
+    def _before_indexed_component(self, visitor, child):
+        visitor.env[visitor.symbolmap.getSymbol(child)] = child
+        return False, (_CONSTANT, child)
+
+    def _before_index_template(self, visitor, child):
+        symb = visitor.symbolmap.getSymbol(child)
+        visitor.env[symb] = 0
+        visitor.expr_cache[id(child)] = child
+        return False, (_CONSTANT, child)
+
+    def _before_component(self, visitor, child):
+        visitor.env[visitor.symbolmap.getSymbol(child)] = child
+        return False, (_CONSTANT, child)
```

Comment:
Similar ansert: this gets called for components in the expression tree that are not indexed, and not `is_numeric_type()` or `is_logical_type()`.  You can get there with something like `sum(m.b[i].c.x for i in m.I)`: `c` is a Block component in the expression tree.

Reply:
OK, this makes sense. But is it safe to always assume that it's constant? Like, what if it was a `PiecewiseLinearExpression` or something else ludicrous?

Reply:
(That might inherit from Block...)

Reply:
I think so - because if the component could be treated like a numeric or logical type, then it should have returned True for either `is_numeric_type()` or `is_logical_type()` (and `BeforeChildDispatcher.register_dispatcher` should have registered a different handler.

Reply:
Ahhhh, okay, that makes sense.